### PR TITLE
[Identity Framework] Add APIs to support federated user account associations

### PR DESCRIPTION
**Pull request by tharindu-bandara**
### Proposed changes in this pull request

- API modifications to support $subject.
- This PR introduces a new OSGi service([FederatedAssociationManager](components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/FederatedAssociationManager.java)) for Federated association management. 
- The existing federated association APIs in the UserProfileAdmin class is refactored to use this newly introduced OSGi service
- Resolves wso2/product-is#6748.

### CODE REVIEW

File: components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/exception/FederatedAssociationManagerClientException.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (c) 2015, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (madurangasiriwardena) [inline components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/exception/FederatedAssociationManagerClientException.java]: License year

File: components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/exception/FederatedAssociationManagerException.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (c) 2015, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (madurangasiriwardena) [inline components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/exception/FederatedAssociationManagerException.java]: License year

File: components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/UserProfileAdmin.java [hunk 3]
```diff
@@ -663,11 +671,11 @@ public void associateID(String idpID, String associatedID) throws UserProfileExc
      * @throws UserProfileException
      */
     public String getNameAssociatedWith(String idpID, String associatedID) throws UserProfileException {
-
+        
         try {
-            return UserProfileMgtDAO.getInstance().getUserAssociatedFor(CarbonContext.getThreadLocalCarbonContext()
-                    .getTenantId(), idpID, associatedID);
-        } catch (UserProfileException e) {
+            return getFederatedAssociationManager().getUserForTheFederatedAssociation(
```
Comment (madurangasiriwardena) [inline components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/UserProfileAdmin.java]: let's remove "The" part

File: components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/FederatedAssociationManagerImpl.java [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+    }
+
+    private boolean isValidFederatedAssociation(String userName, String federatedAssociationId)
+            throws FederatedAssociationManagerException {
+
+        FederatedAssociation[] federatedUserAccountAssociationDTOS
+                = getFederatedAssociationsOfUser(userName);
+        if (federatedUserAccountAssociationDTOS != null) {
+            for (FederatedAssociation federatedUserAccountAssociationDTO
+                    : federatedUserAccountAssociationDTOS) {
```
Comment (madurangasiriwardena) [inline components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/FederatedAssociationManagerImpl.java]: let's get this line to the above line

File: components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/FederatedAssociationManagerImpl.java [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+    private String getUsernameWithoutDomain(String username) {
+
+        int index = username.indexOf(CarbonConstants.DOMAIN_SEPARATOR);
+        if (index < 0) {
+            return username;
+        }
+        return username.substring(index + 1);
+    }
+
+    private String removeTenantDomain(String username) {
```
Comment (madurangasiriwardena) [inline components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/FederatedAssociationManagerImpl.java]: MultitenantUtils.getTenantAwareUsername() will do the same thing

**Comment by malithie** [comment]
@tharindu-bandara ,
Is there any place that I can refer to the feature requirement and high level user story with related to the API.

**Comment by tharindu-bandara** [comment]
Hi @malithie,

The [public issue](https://github.com/wso2/product-is/issues/6748) describes the initial requirement. This pull request caters to the framework changes required, by exposing the [new OSGi service]( https://github.com/wso2/carbon-identity-framework/pull/components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/FederatedAssociationManager.java).

I have added more information to the PR description as well. Please let me know if you require further clarification.

File: components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/FederatedAssociationManagerImpl.java [hunk 6]
```diff
@@ -0,0 +1,498 @@
...
+                        + MultitenantUtils.getTenantDomain(userId));
+            }
+            throw handleFederatedAssociationManagerServerException(ERROR_WHILE_WORKING_WITH_FEDERATED_ASSOCIATIONS, e,
+                    false);
+        }
+        validateTenantId(MultitenantUtils.getTenantDomain(userId), tenantId);
+        return tenantId;
+    }
+
+    private int getIdFromTenantDomain(String tenantDomain) throws FederatedAssociationManagerException {
```
[the remark this answers is not in the document]
Comment (tharindu-bandara) [reply components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/FederatedAssociationManagerImpl.java]: This method already exists to validate and get tenant Id from the user id. I will rename this method's name as `getTenantIdFromTenantDomain`

File: components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/exception/FederatedAssociationManagerClientException.java (hunk 1 above)
Comment (tharindu-bandara) [reply components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/exception/FederatedAssociationManagerClientException.java]: Fixed in 78f08459d2c398277e992c0c738443cfe4edad1b

File: components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/exception/FederatedAssociationManagerException.java (hunk 2 above)
Comment (tharindu-bandara) [reply components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/exception/FederatedAssociationManagerException.java]: Fixed in 78f08459d2c398277e992c0c738443cfe4edad1b

File: components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/UserProfileAdmin.java (hunk 3 above)
Comment (tharindu-bandara) [reply components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/UserProfileAdmin.java]: Fixed in 78f08459d2c398277e992c0c738443cfe4edad1b

File: components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/FederatedAssociationManagerImpl.java (hunk 5 above)
Comment (tharindu-bandara) [reply components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/FederatedAssociationManagerImpl.java]: Fixed in 78f08459d2c398277e992c0c738443cfe4edad1b

File: components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/FederatedAssociationManagerImpl.java (hunk 4 above)
Comment (tharindu-bandara) [reply components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/FederatedAssociationManagerImpl.java]: Fixed in 78f08459d2c398277e992c0c738443cfe4edad1b

File: components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/FederatedAssociationManagerImpl.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (tharindu-bandara) [reply components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/FederatedAssociationManagerImpl.java]: Fixed in 78f08459d2c398277e992c0c738443cfe4edad1b

File: components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/UserProfileAdmin.java [hunk 7]
```diff
@@ -850,60 +846,87 @@ private boolean isAdminProfileSpoof(String username) throws UserStoreException {
...
-                "FederatedID" + "\" : \"" + federatedUserID + "\"";
+        return e.getErrorCode().contains(String.valueOf(INVALID_FEDERATED_ASSOCIATION.getCode()));
     }
 
+    private void validateFederatedAssociationParameters(String username, String idpID, String associatedID,
+                                                        String auditData) throws UserProfileException {
+
+        if (StringUtils.isEmpty(idpID) || StringUtils.isEmpty(associatedID)) {
+            if (log.isDebugEnabled()) {
+                log.debug("Required parameters, idpId and the associatedId are empty.");
```
Comment (IsuraD) [inline components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/UserProfileAdmin.java]: idpId or the associatedId are empty.

File: components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/FederatedAssociationManager.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+            throws FederatedAssociationManagerException;
+
+    /**
+     * Return an array of federated associations associated with the given user.
+     *
+     * @param userId Fully qualified user name of the user.
+     * @return An array of federated associations which contains the federated identifier info.
+     * @throws FederatedAssociationManagerException
+     */
+    FederatedAssociation[] getFederatedAssociationsOfUser(String userId) throws FederatedAssociationManagerException;
```
Comment (IsuraD) [inline components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/FederatedAssociationManager.java]: Lets use user obejct here

Comment (tharindu-bandara) [reply components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/association/federation/FederatedAssociationManager.java]: Fixed in 16a9f7f8559dca533109b50dd6036a753511b309

File: components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/UserProfileAdmin.java (hunk 7 above)
Comment (tharindu-bandara) [reply components/user-mgt/org.wso2.carbon.identity.user.profile/src/main/java/org/wso2/carbon/identity/user/profile/mgt/UserProfileAdmin.java]: Fixed in 16a9f7f8559dca533109b50dd6036a753511b309

**Pull request merged**